### ISSUE TITLE
net: IPNet missing network mask on Windows

### ISSUE THREAD

**Issue by akevinbailey** [golang/go#12811]
Go 1.5.1 on Windows the <code>IPNet</code> structure retuned by <code>net.Interfaces().Addrs()</code> does not return the actual network mask of the IP address.

The current code uses the <code>puni.Address.SockaddrLength</code>, but this only returns the default mast of the network and NOT the actual mask!  Instead you need to use <code>puni.OnLinkPrefixLength</code>

Fix:
In the <code>interfaceAddrTable</code> function in interface_windows.go change:
<code>puni.Address.SockaddrLength</code> to <code>puni.OnLinkPrefixLength</code>
in lines 167, 171, 182, 196

example:
<code>
ifa := &net.IPNet{IP: make(net.IP, net.IPv4len), Mask: net.CIDRMask(int(puni.Address.SockaddrLength), 8 * net.IPv4len)}
</code> 
change to
<code>
ifa := &net.IPNet{IP: make(net.IP, net.IPv4len), Mask: net.CIDRMask(int(puni.OnLinkPrefixLength), 8 * net.IPv4len)}
</code> 

This will only work on Widows Vista/Windows2008 and above.  For Windows XP/Windows 2003 you need to implement the following C code in Go to get the correct prefix length (network mask):

<code>
/* Prior to Windows Vista the FirstPrefix pointed to the list with single prefix for each IP address assigned to the adapter. Order of FirstPrefix does not match order of FirstUnicastAddress, so we need to find corresponding prefix  */

        IP_ADAPTER_PREFIX* prefix;
        prefix_len = 0;

        for (prefix = adapter->FirstPrefix; prefix; prefix = prefix->Next) {
          /* We want the longest matching prefix. */
          if (prefix->Address.lpSockaddr->sa_family != sa->sa_family ||
              prefix->PrefixLength <= prefix_len)
            continue;

          if (address_prefix_match(sa->sa_family, sa,
              prefix->Address.lpSockaddr, prefix->PrefixLength)) {
            prefix_len = prefix->PrefixLength;
          }
        }

        /* If there is no matching prefix information, return a single-host
         * subnet mask (e.g. 255.255.255.255 for IPv4).
         */
        if (!prefix_len)
          prefix_len = (sa->sa_family == AF_INET6) ? 128 : 32;
</code>

I'll leave this exercise up to the Go guys.

**Comment by akevinbailey** [comment]
I found another issue:  The <code>interfaceAddrTable</code> returns Interfaces that are not active.  Need to implement the follow code after for <code>for ; paddr != nil; paddr = paddr.Next {</code>: 

<code>if paddr.OperStatus == windows.IfOperStatusUp {
 ... 
}</code>

This will exclude inactive interfaces.

**Comment by alexbrainman** [comment]
@akevinbailey please review https://go-review.googlesource.com/16854. I just followed your advice. I really don't know much about the subject. Thank you.

Alex

**Comment by rsc** [comment]
Is this a regression from Go 1.4? If not, I think it should be released in Go 1.6, not Go 1.5.2.

**Comment by akevinbailey** [comment]
The network mask code looks good, but I don't have a Windows XP image to test the first prefix code.  However, you left out a way to filter out the non active interfaces.  See https://github.com/golang/go/issues/12812

I think it would be best to merge the fix for this issue (12811) with 12812.  Then test and release them together as part of Go 1.6.

**Comment by alexbrainman** [comment]
> Is this a regression from Go 1.4? ...

It is not regression from go1.4. It was introduced by CL 3024, which is part of go1.5.

> The network mask code looks good, but I don't have a Windows XP image to test the first prefix code.

I have Windows XP computer. I visually checked my new netmasks looks good.

> However, you left out a way to filter out the non active interfaces. See #12812

CL 16854 fixes issue #12811. It has nothing to do with issue #12812. One thing at a time. Do you have a suggestion for a test I can use for issue #12812?

> I think it would be best to merge the fix for this issue (12811) with 12812. Then test and release them together as part of Go 1.6.

I disagee. These bugs are not related in any ways. It will be much easier (and quicker) to have 2 simple fixes approved, then 1 combined - bigger and more complicated. Both fixes can go into go1.6 (if they make deadline).

Alex

**Comment by akevinbailey** [comment]
Several months ago I combined the two "fixes" to get the desired code for my application.  However, since you are the Go expert, you would know better than I what is the best way to bring this to general availability.

I found this issue because my laptop has WiFi, 1000base-T, and several VPN connections as shown in the image below.  All of the interfaces that have an X are not active but show up in <code>paddr</code> list.  The ones that are actually active will have a <code>paddr.OperStatus == windows.IfOperStatusUp</code>.

![networksettings](https://cloud.githubusercontent.com/assets/13941653/11174152/f7be9148-8bd4-11e5-8c0e-c60e32c787ba.png)
